docs/incubator-proposal.md

```diff
@@ -0,0 +1,91 @@
+# cert-manager proposal
+
+## Problem
+
+Currently in Kubernetes, a number of different components including user
+applications require TLS certificates. Right now, users can provide
+certificates to workloads on the cluster by creating `Secret` resources that
+contain a private/public keypair.
+
+A number of projects (see: [kube-lego](https://github.com/jetstack/kube-lego),
+[kube-cert-manager](https://github.com/PalmStoneGames/kube-cert-manager)
+and [kubernetes-letsencrypt](https://github.com/tazjin/kubernetes-letsencrypt))
+came about in order to automate the retrieval of certificates from an [ACME](https://github.com/ietf-wg-acme/acme/)
+compliant server. This has served the community well, especially with the rise
+in popularity of ACME due to [letsencrypt](https://letsencrypt.org). These
+projects also attempt to renew the certificate as it reaches it's expiry time.
+
+Whilst each project has invented it's own schema/methodology for obtaining
+these certificates, the end goal has largely been the same. To provide a signed
+keypair in a Secret resource to then be consumed, either by an Ingress
+controller or some other workload in-cluster, and then keep that keypair up to
+date.
+
+Looking beyond the capabilities of all of these projects, it'd also be
+beneficial for users to be able to utilise alternative issuer backends, other
+than ACME. This could include [Hashicorp Vault](https://vaultproject.io), a
+plain CA (backed by Kubernetes secret resources), or any other cloud based
+service.
+
+It is therefore proposed that a new project, `cert-manager`, is created that
+aims to generalise the issuance and renewal of certificates from an arbitrary
+certificate source. This project should be designed in a similar manner to
+other Kubernetes controllers (eg. kube-controller-manager, service-catalog et
+al.).
+
+## Goals
+
+* Provide a non-opinionated interface for new issuers to be added to
+cert-manager
+
+* Provide reliable support for at least ACME HTTP01 and DNS01 challenge types
+in the ACME provider.
+
+* Provide clear, understable Events/log messages to users to inform them that
```

Comment:
I think one of the goals should be to expose a metrics endpoints for Prometheus etc. that reports number of certificates that cannot be renewed. Alerting in case of TLS is vastly important. I've witnessed large cloud providers falling over when one of their TLS certs have expired. I think alerting for failures for certs in near-expiry period is critical.

I don't think anyone will query events/logs in their spare time –or set up alerting based on those. A lot of people use Prometheus etc, so we should probably just leverage that. Consider making this a short/medium-term goal.